File: server/src/Makc2021.Layer1/Common/CommonModule.cs
// Copyright (c) 2021 Maxim Kuzmin. All rights reserved. Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.Extensions.DependencyInjection;

namespace Makc2021.Layer1.Common
{
    /// <summary>
    /// Общий модуль.
    /// Используется как базовый класс всех модулей - классов,
    /// предназначенных для подготовки к работе различных частей приложения.    
    /// </summary>
    public abstract class CommonModule
    {
        #region Public methods

        /// <summary>
        /// Настроить сервисы.
        /// </summary>
        /// <param name="services">Сервисы.</param>
        public abstract void ConfigureServices(IServiceCollection services);

        /// <summary>
        /// Получить не импортированные типы.
        /// </summary>
        /// <param name="allExports">Все экспортированные типы.</param>
        /// <returns>Не импортированные типы.</returns>
        public IEnumerable<Type> GetNotImportedtTypes(HashSet<Type> allExports)
        {
            return GetImports().Where(x => !allExports.Contains(x));
        }

        /// <summary>
        /// Получить экспортированные типы.
        /// </summary>
        /// <returns>Экспортированные типы.</returns>
        public abstract IEnumerable<Type> GetExports();

        #endregion Public methods

        #region Protected methods

        /// <summary>
        /// Получить импортированные типы.
        /// </summary>
        /// <returns>Импортированные типы.</returns>
        protected abstract IEnumerable<Type> GetImports();

        #endregion Protected methods
    }
}

File: client/makc2021-react-redux-ts/src/Layer1/Common/CommonModule.ts
// Copyright (c) 2021 Maxim Kuzmin. All rights reserved. Licensed under the MIT License.

/**
 * Общий модуль.
 */
export abstract class CommonModule {
  private _isConfigured = false;

  /**
   * Конструктор.
   * @param _moduleName Имя модуля.
   */
  constructor(private _moduleName: string) {}

  protected completeConfiguration() {
    this._isConfigured = true;
  }

  /**
   * Выбросить ошибку, если модуль не настроен.
   */
  throwErrorIfModuleIsNotConfigured() {
    if (!this._isConfigured) {
      throw new Error(`Module ${this._moduleName} is not configured`);
    }
  }
}
